From 0cf7144481ca1a4a71490140a19970ec045118be Mon Sep 17 00:00:00 2001
From: Przemyslaw Czarnota <pczarnota@maxlinear.com>
Date: Wed, 6 Apr 2022 10:42:21 +0200
Subject: bridge: add rtnl_link_bridge_delete()

The new function rtnl_link_bridge_delete(), sends a RTM_DELLINK message
just like rtnl_link_delete(), but it appends additional parameters to
the message, thus not causing the kernel to remove the network device.

Instead, the remove operation is carried out on attributes e. g. we
can remove a VLAN from the bridge port. This will call
ndo_bridge_dellink() in the kernel.
---
 include/netlink-private/route/link/api.h |  4 ++++
 include/netlink/route/link/bridge.h      |  3 +++
 lib/route/link.c                         | 26 ++++++++++++++++++++++++
 lib/route/link/bridge.c                  | 21 +++++++++++++++++++
 libnl-route-3.sym                        |  1 +
 5 files changed, 55 insertions(+)

diff --git a/include/netlink-private/route/link/api.h b/include/netlink-private/route/link/api.h
index 6d30bf7..68a87fc 100644
--- a/include/netlink-private/route/link/api.h
+++ b/include/netlink-private/route/link/api.h
@@ -191,6 +191,10 @@ extern int			rtnl_link_af_data_compare(struct rtnl_link *a,
 extern int			rtnl_link_info_data_compare(struct rtnl_link *a,
 							    struct rtnl_link *b,
 							    int flags);
+extern int			rtnl_link_build_link_request(int,
+							     struct rtnl_link *,
+							     int,
+							     struct nl_msg **);
 
 #ifdef __cplusplus
 }
diff --git a/include/netlink/route/link/bridge.h b/include/netlink/route/link/bridge.h
index 2281a88..487df88 100644
--- a/include/netlink/route/link/bridge.h
+++ b/include/netlink/route/link/bridge.h
@@ -98,6 +98,9 @@ extern int rtnl_link_bridge_set_vlan_range(struct rtnl_link *link, uint16_t vid_
 extern int rtnl_link_bridge_set_master(struct rtnl_link *link);
 
 extern struct rtnl_link_bridge_vlan *rtnl_link_bridge_get_port_vlan(struct rtnl_link *link);
+
+extern int rtnl_link_bridge_delete(struct nl_sock *sk,
+				   struct rtnl_link *bridge);
 #ifdef __cplusplus
 }
 #endif
diff --git a/lib/route/link.c b/lib/route/link.c
index 260185b..5bd8b3f 100644
--- a/lib/route/link.c
+++ b/lib/route/link.c
@@ -1704,6 +1704,32 @@ int rtnl_link_build_add_request(struct rtnl_link *link, int flags,
 	return build_link_msg(RTM_NEWLINK, &ifi, link, flags, result);
 }
 
+/**
+ * Build a netlink message requesting operation on the virtual link
+ * @arg link		link
+ * @arg flags		additional netlink message flags
+ * @arg result		pointer to store resulting netlink message
+ *
+ * The behaviour of this function is identical to rtnl_link_build_add_request()
+ * with the exception that it allows to specify message type
+ *
+ * @see rtnl_build_add_request()
+ *
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_link_build_link_request(int cmd, struct rtnl_link *link, int flags,
+				 struct nl_msg **result)
+{
+	struct ifinfomsg ifi = {
+		.ifi_family = link->l_family,
+		.ifi_index = link->l_index,
+		.ifi_flags = link->l_flags,
+		.ifi_change = link->l_flag_mask,
+	};
+
+	return build_link_msg(cmd, &ifi, link, flags, result);
+}
+
 /**
  * Add virtual link
  * @arg sk		netlink socket.
diff --git a/lib/route/link/bridge.c b/lib/route/link/bridge.c
index 1e8d37d..03e49c9 100644
--- a/lib/route/link/bridge.c
+++ b/lib/route/link/bridge.c
@@ -609,6 +609,27 @@ int rtnl_link_bridge_add(struct nl_sock *sk, const char *name)
 	return err;
 }
 
+/**
+ * Send a RTM_DELLINK netlink message
+ * @arg sk              netlink socket
+ * @arg bridge          link object
+ *
+ * @return 0 on success or a negative error code
+*/
+int rtnl_link_bridge_delete(struct nl_sock *sk, struct rtnl_link *link)
+{
+	struct nl_msg *msg;
+	int err;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	err = rtnl_link_build_link_request(RTM_DELLINK, link, 0, &msg);
+	if (err < 0)
+		return err;
+
+	return nl_send_sync(sk, msg);
+}
+
 /**
  * Check if a link is a bridge
  * @arg link		Link object
diff --git a/libnl-route-3.sym b/libnl-route-3.sym
index 7ed00d7..ca4b091 100644
--- a/libnl-route-3.sym
+++ b/libnl-route-3.sym
@@ -261,6 +261,7 @@ global:
 	rtnl_link_bond_release_ifindex;
 	rtnl_link_bridge_add;
 	rtnl_link_bridge_alloc;
+	rtnl_link_bridge_delete;
 	rtnl_link_bridge_flags2str;
 	rtnl_link_bridge_get_cost;
 	rtnl_link_bridge_get_flags;
